src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,235 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
+
+namespace Microsoft.CodeAnalysis.IntroduceUsingStatement
+{
+    internal abstract class AbstractIntroduceUsingStatementCodeRefactoringProvider<TStatementSyntax, TLocalDeclarationSyntax, TBlockSyntax> : CodeRefactoringProvider
+        where TStatementSyntax : SyntaxNode
+        where TLocalDeclarationSyntax : TStatementSyntax
+        where TBlockSyntax : SyntaxNode
+    {
+        protected abstract SyntaxList<TStatementSyntax> GetStatements(TBlockSyntax blockSyntax);
+
+        protected abstract TBlockSyntax WithStatements(TBlockSyntax blockSyntax, SyntaxList<TStatementSyntax> statements);
+
+        protected abstract TStatementSyntax CreateUsingStatement(TLocalDeclarationSyntax declarationStatement, SyntaxList<TStatementSyntax> statementsToSurround);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var span = context.Span;
+
+            var (declarationSyntax, _) =
+                await FindDisposableLocalDeclaration(document, span, context.CancellationToken).ConfigureAwait(false);
+
+            if (declarationSyntax != null)
+            {
+                context.RegisterRefactoring(new MyCodeAction(
+                    "Introduce using statement",
+                    cancellationToken => IntroduceUsingStatementAsync(document, span, cancellationToken)));
+            }
+        }
+
+        private static async Task<(TLocalDeclarationSyntax, ILocalSymbol)> FindDisposableLocalDeclaration(Document document, TextSpan selection, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var declarationSyntax = root.FindNode(selection)?.FirstAncestorOrSelf<TLocalDeclarationSyntax>();
```

Comment:
you can just do.  FindToken(selection.start).FirstAncestor.  No need for FindNode, or ?.

Reply:
Actually, never mind. I can't figure out how to get the declaration from this selection with `FindToken*`, and `FindNode` gets it right:

```c#
var x = disposable;[||]
Unrelated();
```